### PR TITLE
Fixes #16860: Improving exception experience

### LINKED ISSUE #16860: Improve exception experience
For details see https://github.com/OmniSharp/omnisharp-vscode/issues/1007

We could consider to revive the Peek UI.

### PR DESCRIPTION
**Pull request by michelkaporin**
I have added an exception widget that shows a thrown exception message in a debug mode when such occurs.

image

Fixes #16860.

Also removed border colours from widget options and moved it to CSS. Pointer arrow for widget colour moved also to CSS from the code.

### CODE REVIEW

File: src/vs/editor/contrib/gotoError/browser/gotoError.css [hunk 1]
```diff
@@ -12,16 +13,28 @@
...
 }
 
+
 /* High Contrast Theming */
+
 .monaco-editor.hc-black .marker-widget {
 	background-color: #0C141F;
 }
 
-.monaco-editor .marker-widget > .stale {
```
Comment (michelkaporin) [inline src/vs/editor/contrib/gotoError/browser/gotoError.css]: Do not change unnecesery lines

File: src/vs/editor/contrib/gotoError/browser/gotoError.ts [hunk 2]
```diff
@@ -269,11 +269,11 @@ class MarkerNavigationWidget extends ZoneWidget {
 			// update frame color (only applied on 'show')
 			switch (marker.severity) {
 				case Severity.Error:
-					this.options.frameColor = '#ff5a5a';
+					this._container.parentElement.classList.add('marker-error-widget');
```
Comment (michelkaporin) [inline src/vs/editor/contrib/gotoError/browser/gotoError.ts]: Make sure to remove the warning class in this case, same as when you add a warning class remove the error class

File: src/vs/editor/contrib/zoneWidget/browser/zoneWidget.css [hunk 3]
```diff
@@ -16,15 +16,15 @@
 }
```
Comment (michelkaporin) [inline src/vs/editor/contrib/zoneWidget/browser/zoneWidget.css]: Looks good

File: src/vs/editor/contrib/zoneWidget/browser/zoneWidget.ts [hunk 4]
```diff
@@ -18,7 +18,6 @@ import { ICodeEditor, IOverlayWidget, IOverlayWidgetPosition, IViewZone, IViewZo
 export interface IOptions {
 	showFrame?: boolean;
```
Comment (michelkaporin) [inline src/vs/editor/contrib/zoneWidget/browser/zoneWidget.ts]: Make sure that references widget, peek view widget, error list widget and breakpoint widget are rendered the same as before.

File: src/vs/editor/contrib/gotoError/browser/gotoError.css [hunk 5]
```diff
@@ -2,6 +2,7 @@
  *  Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (michelkaporin) [inline src/vs/editor/contrib/gotoError/browser/gotoError.css]: Looks good

File: src/vs/workbench/parts/debug/browser/media/breakpointWidget.css [hunk 6]
```diff
@@ -6,6 +6,8 @@
 .monaco-editor .zone-widget .zone-widget-container.breakpoint-widget {
 	height: 30px !important;
 	display: flex;
```
Comment (michelkaporin) [inline src/vs/workbench/parts/debug/browser/media/breakpointWidget.css]: Try to put it in one rule together

File: src/vs/workbench/parts/debug/browser/media/debug.contribution.css [hunk 7]
```diff
@@ -8,14 +8,12 @@
 	background-image: url('debug.svg');
 }
 
-.monaco-editor .debug-top-stack-frame-line {
+.monaco-editor .debug-top-stack-frame-line,
+.monaco-editor .debug-top-stack-frame-exception-line
+{
```
Comment (michelkaporin) [inline src/vs/workbench/parts/debug/browser/media/debug.contribution.css]: Put the bracket on the previous line

File: src/vs/workbench/parts/debug/electron-browser/debugEditorContribution.ts [hunk 8]
```diff
@@ -296,6 +305,47 @@ export class DebugEditorContribution implements IDebugEditorContribution {
 		}
 	}
 
+	// exception widget
```
Comment (michelkaporin) [inline src/vs/workbench/parts/debug/electron-browser/debugEditorContribution.ts]: Why is this public?

File: src/vs/workbench/parts/debug/electron-browser/debugEditorContribution.ts [hunk 9]
```diff
@@ -296,6 +305,47 @@ export class DebugEditorContribution implements IDebugEditorContribution {
...
+	private showExceptionWidget(lineNumber: number): void {
+		if (this.exceptionWidget) {
+			this.exceptionWidget.dispose();
+		}
+
+		this.exceptionWidget = this.instantiationService.createInstance(ExceptionWidget, this.editor, lineNumber);
+		this.exceptionWidget.show({ lineNumber, column: 1 }, 3);
+	}
+
+	private hideExceptionWidget() {
```
Comment (michelkaporin) [inline src/vs/workbench/parts/debug/electron-browser/debugEditorContribution.ts]: Get rid fo this method

File: src/vs/workbench/parts/debug/electron-browser/debugEditorContribution.ts [hunk 10]
```diff
@@ -296,6 +305,47 @@ export class DebugEditorContribution implements IDebugEditorContribution {
...
+		this.exceptionWidget.hide();
+	}
+
+	private closeExceptionWidget(): void {
```
Comment (michelkaporin) [inline src/vs/workbench/parts/debug/electron-browser/debugEditorContribution.ts]: No need to make difference between hide and close. Always destroy and recreate the widget as needed

File: src/vs/workbench/parts/debug/electron-browser/debugEditorContribution.ts [hunk 11]
```diff
@@ -296,6 +305,47 @@ export class DebugEditorContribution implements IDebugEditorContribution {
...
+		// Toggles exception widget based on the state of the current editor model and debug stack frame
+		const model = this.editor.getModel();
+		const focusedSf = this.debugService.getViewModel().focusedStackFrame;
+		const callStack = focusedSf ? focusedSf.thread.getCallStack() : null;
+		if (!model || !focusedSf || !callStack) {
+			this.closeExceptionWidget();
+			return;
+		}
+
+		// First call stack frame is the frame where exception has been thrown
```
Comment (michelkaporin) [inline src/vs/workbench/parts/debug/electron-browser/debugEditorContribution.ts]: Callstack can be an empty array in some corner cases

**Comment by isidorn** [comment]
@michelkaporin looks good! Let's just do one more thing before we merge it in. The exception widget needs to respect the column of the top stack frame. Namely it shoudl have an arrow just like the find references widget pointing to the exact column location.
Once that is done we can merge this PR in - thanks!

**Comment by michelkaporin** [comment]
@isidorn done. I've also noticed the background colour was applied to the arrow because it was inherited from zone-widget.css due to CSS refactoring. Please see 706773bb346e4a19c5bd56c2e902c8fe80e80b4c for this.

**Comment by isidorn** [comment]
@michelkaporin looks good, merging in. I will still test it on my machine a bit and will drop by your machine so we polish it more.

**Pull request merged**